Adding account key roll-over

felixfontein opened the pull request:
Adds [ACME-13](https://tools.ietf.org/html/draft-ietf-acme-acme-13#section-7.3.6) account key roll-over to Pebble.

Fixes #150.

felixfontein commented:
Ok, there's a major downside with this: currently, account IDs are computed from the account key, so a key change also changes the account ID. This is not a good idea, I think, and clients storing the account ID will have problems with this.

Idea:
 1. Put this PR on hold;
 2. Create another PR to detach account IDs from keys.

For 2., the account ID should probably be some random string. I guess the DB should do the mapping between keys and account IDs. Is that OK? If yes, I can start working on that.

Another (minor) point with this PR: it doesn't check whether the new key is already used for an account.

cpu commented:
:wave: Hi @felixfontein, thanks for the PR & accompanying thoughts.

> Ok, there's a major downside with this: currently, account IDs are computed from the account key, so a key change also changes the account ID. This is not a good idea, I think, and clients storing the account ID will have problems with this.

Hmmm. That is a good point! I don't think we had this consequence in mind when we made the choice to identify accounts this way.

> For 2., the account ID should probably be some random string. I guess the DB should do the mapping between keys and account IDs. Is that OK? If yes, I can start working on that.

Yes, this sounds good. I'm in agreement with you that holding out on this PR until the related account ID problem is fixed is a good course of action. Thanks so much for offering to work on a fix :-)

> Another (minor) point with this PR: it doesn't check whether the new key is already used for an account.

I think that's probably worth addressing before we land this code since it could end up surprising folks and performing this check is a MUST for server operators in the draft specification.

felixfontein commented:
@cpu I rebased against a squashed version of #154 (Using account IDs different from key IDs) and adjusted the PR to behave correctly ;) I added the missing existing account public key check, and also a check for the URL parameters of the inner and outer JWS being the same. I think the PR now behaves as expected :)

As soon as #154 is merged, I'll rebase against the new master to get rid of the commits from #154.

cpu commented:
> As soon as #154 is merged, I'll rebase against the new master to get rid of the commits from #154.

@felixfontein All set for :point_up: 

I'll try to give this a review pass today but it might slip until next week. Thanks again!

felixfontein commented:
@cpu Great! I've just rebased it to current `master`, so it's ready for review :)

cpu commented on db/memorystore.go:
```diff
@@ -116,6 +119,31 @@ func (m *MemoryStore) AddAccount(acct *core.Account) (int, error) {
 	return len(m.accountsByID), nil
 }
 
+func (m *MemoryStore) ChangeAccountKey(acct *core.Account, newKey *jose.JSONWebKey) (error, *core.Account) {
```
Can you swap the order of the `*core.Account` and `error` returns? It's common Go convention to always return the `error` type as the last return value.

cpu replied:
Re-reading this I understand now that the `*core.Account` return was being used for key-conflicts. I think you should remove this return entirely and just return an `error` and use the `existingAccountError` I describe in my other comment for that case.

cpu commented on db/memorystore.go:
```diff
@@ -116,6 +119,31 @@ func (m *MemoryStore) AddAccount(acct *core.Account) (int, error) {
+	acct.Key = newKey
+	m.accountsByKeyID[newKeyID] = acct
+	m.accountsByID[acct.ID] = acct
+	return nil, nil
```
I think this should be just `return nil` once the refactoring I describe happens.

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' required for inner JWS."), response)
+	}
+	if innerHeaderURL != outerHeaderURL {
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' differ for inner and outer JWS."), response)
```
nit: I think "differ" should be plural, e.g. "JWS header parameter 'url' differs for inner and outer JWS."

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+	}
+
+	var innerContent struct {
+		Account string `json:"account"`
```
You can get rid of the `json` tags on both of these fields since their name matches the JSON keys exactly and you aren't specifying any additional options in the tags.

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+	}
+	err = json.Unmarshal([]byte(innerPayload), &innerContent)
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem("Error unmarshaling key roll-over innermost JSON body"), response)
```
I think we use the term "inner JWS body" in Boulder. Can you use that here as well instead of "innermost JSON body"?

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+	// Check account ID
+	prefix := wfe.relativeEndpoint(request, acctPath)
+	if !strings.HasPrefix(innerContent.Account, prefix) {
+		wfe.sendError(acme.MalformedProblem("Key ID (account) in innermost JSON body missing expected URL prefix"), response)
```
ditto: re innermost JSON body

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+	}
+	accountID := strings.TrimPrefix(innerContent.Account, prefix)
+	if accountID == "" {
+		wfe.sendError(acme.MalformedProblem("No key ID (account) in innermost JSON body"), response)
```
Can you include the raw `innerContent.Account` as a `%q` format specifier somewhere in the error message? That will make debugging easier for client developers.

cpu replied:
(also another place to fix "innermost JSON body")

felixfontein replied:
Done.

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+		return
+	}
+	if accountID != existingAcct.ID {
+		wfe.sendError(acme.MalformedProblem("Key roll-over innermost JSON contains wrong account ID"), response)
```
Can you include the raw innerContent.Account as a %q format specifier somewhere in the error message here too? (Also another "innermost JSON" to replace)

felixfontein replied:
Done.

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+
+	// Verify inner key
+	if !keyDigestEquals(innerContent.OldKey, *existingAcct.Key) {
+		wfe.sendError(acme.MalformedProblem("Key roll-over innermost JSON contains wrong old key"), response)
```
another "innermost JSON" to fix.

felixfontein replied:
Done.

cpu commented on db/memorystore.go:
```diff
@@ -116,6 +119,31 @@ func (m *MemoryStore) AddAccount(acct *core.Account) (int, error) {
+	}
+
+	if otherAccount, present := m.accountsByKeyID[newKeyID]; present {
+		return fmt.Errorf("New public key is already in use"), otherAccount
```
In general it's not good Go practice to return both an error and an instance of the non-error return type. 

It would be better here to return just an `error` from this function. Return only `nil` when everything goes OK and return `existingAccountError{otherAccount.ID}` for a key conflict. You would then define your own `existingAccountError` type that implements the error interface, holds an account ID, and prints a good message about there being an existing account with the given ID.

In the calling code you can specifically check whether `err` is of type `existingAccountError`, and if it is, cast it to that type and extract the existing account ID to return as a `Location` header.

It's a bit more work but I think it will be a cleaner design overall

felixfontein replied:
I guess it must be `ExistingAccountError`, otherwise I cannot use it both in `db` and `wfe`.

cpu replied:
@felixfontein Yup! You're correct. It should be an exported type for that reason. Sorry I missed that :-)

felixfontein replied:
@cpu Great! I'm learning a bit more of Go every time I work on this ;)

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +715,102 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+
+	// Ok, now change account key
+	err, otherAccount := wfe.db.ChangeAccountKey(existingAcct, newPubKey)
+	if err != nil {
```
This is where you will want to check whether `err` is of the specific type that you defined for a key conflict, and handle that accordingly.

felixfontein commented:
@cpu Thanks for your review! I've now refactored the key already in use error as you suggested. The next step will be shortening the `wfe.KeyRollover` function.

felixfontein commented:
@cpu I think I'm done refactoring. Can you check how close this is to what you want? :)

cpu commented on db/memorystore.go:
```diff
@@ -16,6 +16,17 @@ import (
+}
+
+func (e *ExistingAccountError) Error() string {
+	return e.msg
```
I think you can remove the `msg` field from this struct entirely since the only place `ExistingAccountError` is used is for a new key conflict and the message for that case is always the same. If you remove `msg` you can implement `Error()` with: 
`return fmt.Sprintf("New key is already in use by account %s", e.ID)`

felixfontein replied:
Done.

cpu commented on db/memorystore.go:
```diff
@@ -16,6 +16,17 @@ import (
 	"github.com/letsencrypt/pebble/core"
 )
 
+// Error when key change fails because new account key has already been in use
```
The Godoc convention is to start a type's comment with the name of the type, like:
> // ExistingAccountError is an error type indicating when an operation fails because the `MatchingAccount` has a key conflict.

felixfontein replied:
Done.

cpu commented on db/memorystore.go:
```diff
@@ -116,6 +130,31 @@ func (m *MemoryStore) AddAccount(acct *core.Account) (int, error) {
+	}
+
+	if otherAccount, present := m.accountsByKeyID[newKeyID]; present {
+		return &ExistingAccountError{"New public key is already in use", otherAccount}
```
If you refactor the `ExistingAccountError` as suggested above this will end up just being `return ExistingAccountError{otherAccount}` (I believe you don't need to be using `&{}` here to take a reference)

felixfontein replied:
The `&` is needed, otherwise go complains with:
```
db/memorystore.go:148:30: cannot use ExistingAccountError literal (type ExistingAccountError) as type error in return argument:
	ExistingAccountError does not implement error (Error method has pointer receiver)
```
I had to replace
```
func (e *ExistingAccountError) Error() string { ...
```
by
```
func (e ExistingAccountError) Error() string { ...
```
then it worked :)

cpu commented on wfe/wfe.go:
```diff
@@ -499,64 +502,80 @@ func (wfe *WebFrontEndImpl) verifyPOST(
 	pubKey *jose.JSONWebKey,
 	parsedJWS *jose.JSONWebSignature,
-	request *http.Request) ([]byte, *jose.JSONWebKey, *acme.ProblemDetails) {
+	jwsPrefix string) ([]byte, string, *acme.ProblemDetails) {
```
Are you sure that you need the `jwsPrefix` argument that you're adding to this function? It looks like `verifyJWS` always sets it to `""` and only `wfe.KeyRollover` uses it. I think it would be better to remove this argument and have `wfe.KeyRollover` make a new malformed problem with the correct message prefix when there is a non-nil err response from calling `verifyJWSLight` (or whatever it ends up being called!)

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +731,109 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+		Account string
+		OldKey  jose.JSONWebKey
+	}
+	err := json.Unmarshal([]byte(innerPayload), &innerContent)
```
I don't think you need the `[]byte()` cast on `innerPayload` since its already a `[]byte` argument

felixfontein replied:
Yes, it also works without it.

cpu commented on wfe/wfe.go:
```diff
@@ -712,6 +731,109 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
+	}
+
+	if innerHeaderURL != outerHeaderURL {
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' differs for inner and outer JWS."), response)
```
I think you missed a `return` after this `wfe.sendError` and control flow will continue when the inner header and outer header don't match

felixfontein replied:
Definitely! Thanks for noticing this!

cpu commented on wfe/wfe.go:
```diff
@@ -499,64 +502,80 @@ func (wfe *WebFrontEndImpl) verifyPOST(
-func (wfe *WebFrontEndImpl) verifyJWS(
+// Checks parsed JWS whether it matches the given public key, does some general
+// checks, and extracts the URL header parameter.
+func (wfe *WebFrontEndImpl) verifyJWSLight(
```
Sorry, I think I forgot to hit submit to include this comment in my last review:

Can we brainstorm a new name for this function? What about splitting it up into two functions named `verifyJWSSignature` and `extractJWSURL`?  I think that will be more descriptive than `verifyJWSLight`.

felixfontein replied:
Yes, please :) I tentatively named it `verifyJWSLight` while I wasn't sure what exactly will end up in it, and since it is a subset of what `verifyJWS` is doing ;-)

felixfontein replied:
I think I'll name the first one `verifyJWSSignatureAndAlgorithm` (at least for now), as a reminder that the code for checking the algorithm is still missing ;-)

felixfontein replied:
I did the refactoring in 15befc4c28430080ad0ef5dc8d825b54f48cf529. I think it's now much better than before.

felixfontein commented:
@cpu Since the commit list is getting longer and longer, do you want me to squash the existing ones into one, or a small number of commits?

felixfontein commented:
(I see you're already squashing when merging, but if that helps you reviewing, I can already squash them before.)

cpu commented:
> @cpu Since the commit list is getting longer and longer, do you want me to squash the existing ones into one, or a small number of commits?

@felixfontein That's ok :-) I'm happy to squash on merge. I (lightly) prefer having the individual commits from the review process.
